src/submodule.c

```diff
@@ -1915,32 +1919,34 @@ static int submodule_load_from_wd_lite(git_submodule *sm)
 }
 
 /**
- * Returns a snapshot of $WORK_TREE/.gitmodules.
+ * Requests a snapshot of $WORK_TREE/.gitmodules.
  *
- * We ignore any errors and just pretend the file isn't there.
+ * Returns GIT_ENOTFOUND in case no .gitmodules file exist
  */
-static git_config *gitmodules_snapshot(git_repository *repo)
+static int gitmodules_snapshot(git_config **snap, git_repository *repo)
 {
 	const char *workdir = git_repository_workdir(repo);
-	git_config *mods = NULL, *snap = NULL;
+	git_config *mods = NULL;
 	git_buf path = GIT_BUF_INIT;
+	int error;
 
-	if (workdir != NULL) {
-		if (git_buf_joinpath(&path, workdir, GIT_MODULES_FILE) != 0)
-			return NULL;
+	if (!workdir)
+		return GIT_ENOTFOUND;
 
-		if (git_config_open_ondisk(&mods, path.ptr) < 0)
-			mods = NULL;
-	}
+	if (git_buf_joinpath(&path, workdir, GIT_MODULES_FILE) != 0)
+		return -1;
+
+	error = git_config_open_ondisk(&mods, path.ptr);
 
 	git_buf_free(&path);
 
 	if (mods) {
-		git_config_snapshot(&snap, mods);
+		git_config_snapshot(snap, mods);
 		git_config_free(mods);
+		return 0;
```

Comment:
We usually try to avoid returning early for non-trivial code. I'd rather go with something like

```
...
*snap = NULL;
...

if ((error = git_config_open_ondisk(...)) < 0)
    goto out;
if (!mods) {
    error = -1;
    goto out;
}

if ((error = git_config_snapshot(mods, config)) < 0)
    goto out;

out:
    git_config_free(mods);
    if (error)
        git_config_free(*snap);
    return error;
```

Reply:
`goto` is better? - Shall I change all the code for this?

Reply:
Sorry :sweat_smile: Seeing that your code already improves the function by quite a lot, I'm satisfied if you check the return code of `git_config_snapshot`, without converting the function to `goto` style.